File: 17-LetterCombinationsOfAPhoneNumber/17-LetterCombinationsOfAPhoneNumber.java
// Last updated: 20/07/2026, 16:24:59
import java.util.ArrayList;
import java.util.List;

class Solution {
    private static final String[] PHONE_MAP = {
        "",     
        "",     
        "abc",  
        "def",  
        "ghi",  
        "jkl",  
        "mno",  
        "pqrs", 
        "tuv",  
        "wxyz"  
    };

    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        
        if (digits == null || digits.length() == 0) {
            return result;
        }
        
        backtrack(digits, 0, new StringBuilder(), result);
        return result;
    }

    private void backtrack(String digits, int index, StringBuilder current, List<String> result) {
        if (index == digits.length()) {
            result.add(current.toString());
            return;
        }

        String letters = PHONE_MAP[digits.charAt(index) - '0'];

        for (char c : letters.toCharArray()) {
            current.append(c);                                  
            backtrack(digits, index + 1, current, result);     
            current.deleteCharAt(current.length() - 1);         
        }
    }
}

File: 17-LetterCombinationsOfAPhoneNumber/17-LetterCombinationsOfAPhoneNumber.py
# Last updated: 20/07/2026, 10:34:50
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        
        if not digits:
            return []
        
        phone = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }
        
        result = []
        
        def backtrack(index, path):
            # If combination is complete
            if len(path) == len(digits):
                result.append(path)
                return
            
            # Get letters for current digit
            letters = phone[digits[index]]
            
            for letter in letters:
                backtrack(index + 1, path + letter)
        
        backtrack(0, "")
        return result